File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/page/image/ImageComponent.ts
module api.content.page.image {

    import Region = api.content.page.region.Region;
    import Form = api.form.Form;
    import FormBuilder = api.form.FormBuilder;
    import OccurrencesBuilder = api.form.OccurrencesBuilder;
    import TextLine = api.form.inputtype.text.TextLine;
    import TextArea = api.form.inputtype.text.TextArea;
    import PropertyTree = api.data.PropertyTree;
    import PropertyIdProvider = api.data.PropertyIdProvider;

    export class ImageComponent extends api.content.page.Component implements api.Equitable, api.Cloneable {

        private image: api.content.ContentId;

        private config: PropertyTree;

        private form: Form;

        constructor(builder?: ImageComponentBuilder) {
            super(builder);
            if (builder) {
                this.image = builder.image;
                this.config = builder.config;
            }
            var formBuilder = new FormBuilder();
            formBuilder.addFormItem(new api.form.InputBuilder().
                setName("caption").
                setInputType(TextArea.getName()).
                setLabel("Caption").
                setOccurrences(new OccurrencesBuilder().setMinimum(0).setMaximum(1).build()).
                build());
            this.form = formBuilder.build();
        }

        getImage(): api.content.ContentId {
            return this.image;
        }

        getForm(): api.form.Form {
            return this.form;
        }

        getConfig(): PropertyTree {
            return this.config;
        }

        setImage(value: api.content.ContentId) {
            this.image = value;
        }

        reset() {
            this.image = null;
            //this.config.                                                                                              Arn Jørund
        }


        toJson(): api.content.page.ComponentTypeWrapperJson {

            var json: ImageComponentJson = <ImageComponentJson>super.toComponentJson();
            json.image = this.image != null ? this.image.toString() : null;
            json.config = this.config != null ? this.config.toJson() : null;

            return <api.content.page.ComponentTypeWrapperJson> {
                ImageComponent: json
            };
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, ImageComponent)) {
                return false;
            }

            var other = <ImageComponent>o;

            if (!super.equals(o)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.image, other.image)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.config, other.config)) {
                return false;
            }

            return true;
        }

        clone(generateNewPropertyIds: boolean = false): ImageComponent {
            return new ImageComponentBuilder(this, generateNewPropertyIds).build();
        }
    }

    export class ImageComponentBuilder extends api.content.page.ComponentBuilder<ImageComponent> {

        image: api.content.ContentId;

        config: PropertyTree;

        constructor(source?: ImageComponent, generateNewPropertyIds: boolean = false) {
            super(source);
            if (source) {
                this.image = source.getImage();
                this.config = source.getConfig().copy(generateNewPropertyIds);
            }
        }

        public setImage(value: api.content.ContentId): ImageComponentBuilder {
            this.image = value;
            return this;
        }

        public setConfig(value: PropertyTree): ImageComponentBuilder {
            this.config = value;
            return this;
        }

        public fromJson(json: ImageComponentJson, region: Region, propertyIdProvider: PropertyIdProvider): ImageComponentBuilder {

            if (json.image) {
                this.setImage(new api.content.ContentId(json.image));
            }

            this.setName(new api.content.page.ComponentName(json.name));


            this.setConfig(PropertyTree.fromJson(json.config, propertyIdProvider));
            this.setParent(region);

            return this;
        }

        public build(): ImageComponent {
            return new ImageComponent(this);
        }
    }
}

File: modules/core-api/src/main/java/com/enonic/xp/content/page/region/ImageComponent.java
package com.enonic.xp.content.page.region;


import java.util.Objects;

import org.apache.commons.lang.StringUtils;

import com.enonic.xp.content.ContentId;
import com.enonic.xp.data.PropertyTree;

public class ImageComponent
    extends Component
{
    private static final String CAPTION = "caption";

    private ContentId image;

    private PropertyTree config;

    public ImageComponent( final Builder builder )
    {
        super( builder );
        this.image = builder.image;
        this.config = builder.config != null ? builder.config : new PropertyTree();
    }

    public static Builder newImageComponent()
    {
        return new Builder();
    }

    public static Builder newImageComponent( final ImageComponent source )
    {
        return new Builder( source );
    }

    @Override
    public Component copy()
    {
        return newImageComponent( this ).build();
    }

    @Override
    public ComponentType getType()
    {
        return ImageComponentType.INSTANCE;
    }

    public ContentId getImage()
    {
        return image;
    }

    public boolean hasConfig()
    {
        return config != null;
    }

    public PropertyTree getConfig()
    {
        return this.config;
    }

    public boolean hasCaption()
    {
        return config.hasProperty( CAPTION ) && StringUtils.isNotBlank( config.getString( CAPTION ) );
    }

    /**
     * Returns value of property "caption" in config.
     */
    public String getCaption()
    {
        return config.getString( CAPTION );
    }


    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }
        if ( !super.equals( o ) )
        {
            return false;
        }

        final ImageComponent that = (ImageComponent) o;

        return Objects.equals( image, that.image ) && Objects.equals( config, that.config );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( super.hashCode(), image, config );
    }

    public static class Builder
        extends Component.Builder
    {
        private ContentId image;

        private PropertyTree config;

        private Builder()
        {
            // Default
        }

        private Builder( final ImageComponent source )
        {
            super( source );
            image = source.image;
            config = source.config != null ? source.config.copy() : null;
        }

        public Builder image( final ContentId value )
        {
            this.image = value;
            return this;
        }

        @Override
        public Builder name( ComponentName value )
        {
            this.name = value;
            return this;
        }

        public Builder name( String value )
        {
            this.name = value != null ? new ComponentName( value ) : null;
            return this;
        }

        public Builder config( final PropertyTree config )
        {
            this.config = config;
            return this;
        }

        public ImageComponent build()
        {
            return new ImageComponent( this );
        }
    }
}
